src/charm.py

```diff
@@ -1,29 +1,130 @@
 #!/usr/bin/env python3
 # Copyright 2021 Canonical Ltd.
 # See LICENSE file for licensing details.
+import json
+import logging
 import textwrap
+import urllib.parse
+from typing import Any, Dict, List, Tuple, Optional
 
+import kubernetes
+import ops
+import requests
+import yaml
 from charms.alertmanager_k8s.v0.alertmanager import AlertmanagerProvider
 from charms.karma_k8s.v0.karma import KarmaConsumer
-import utils
-
-import ops
-from ops.charm import CharmBase, ActionEvent
-from ops.main import main
+from ops.charm import ActionEvent, CharmBase
 from ops.framework import StoredState
-from ops.model import ActiveStatus, MaintenanceStatus, BlockedStatus
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
 from ops.pebble import ChangeError
 
-import urllib.parse
-import requests
-import yaml
-import json
-import logging
-from typing import Dict, List, Optional, Any
+import utils
 
 logger = logging.getLogger(__name__)
 
 
+class K8sServicePatch:
+    @staticmethod
+    def namespace() -> str:
+        """Read the Kubernetes namespace we're deployed in from the mounted service token
+
+        Returns:
+            str: The current Kubernetes namespace
+        """
+        with open("/var/run/secrets/kubernetes.io/serviceaccount/namespace", "r") as f:
+            return f.read().strip()
+
+    @staticmethod
+    def _k8s_auth() -> bool:
+        """Authenticate with the Kubernetes API using an in-cluster service token
+
+        Returns:
+            bool: True represents a successful authentication against the Kubernetes API
+        """
+        # Authenticate against the Kubernetes API using a mounted ServiceAccount token
+        kubernetes.config.load_incluster_config()
+        # Test the service account we've got for sufficient perms
+        api = kubernetes.client.CoreV1Api(kubernetes.client.ApiClient())
+
+        try:
+            api.list_namespaced_service(namespace=K8sServicePatch.namespace())
+        except kubernetes.client.exceptions.ApiException as e:
+            if e.status == 403:
+                # If we can't read a cluster role, we don't have enough permissions
+                # Admin would need to run `juju trust` on this application to continue
+                # TODO inform caller that `juju trust` is needed
+                return False
```

Comment:
What about logging this with a `.warning()`?

Reply:
Replaced retvals with exceptions.
Trying to avoid using the `logging` module in "generic" classes.
Please re-review.